File: packages/proof/src/stepped.ts
// Stepped sims: provable simulations that split into small jobs -- one per (entity, window) --
// so a race of twenty cars over five laps is a hundred proofs a GPU clears in seconds, then
// one aggregation, instead of one proof of everything.
//
// The rule that makes it sound: a tick is SIMULTANEOUS. Every entity's state at tick t is a pure
// function of the whole field at t-1 (`step(ctx, prev, i, t)` may read any entity's previous
// state, never a current one). Then a job for entity i over a window re-derives i's states from
// the witnessed field and checks them against the witness; the aggregation checks every job of
// a window witnessed the SAME field, that each entity's computed trace equals the trace everyone
// witnessed for it, and that windows chain. The TypeScript here is the reference both guest
// programs (the job, the aggregator) are ported from.
//
//   const S = defineStepped({ init, step, done, encode, decode });
//   const run = runStepped(S, input);                 // the whole trajectory, fast
//   const jobs = jobsOf(S, run, span);                // (entity, window) witnesses
//   const outs = jobs.map((j) => proveJob(S, run.ctx, j));   // what each job commits (reference)
//   aggregate(S, input, run.ctx, outs);               // what the aggregator checks and derives

import { sha256 } from "@keel-engine/codec";
import { createPacker, equalBytes } from "./bytes.ts";

export interface SteppedSpec<I, C, E> {
  /** The immutable per-match context (track, stats, keys...) derived from the input. */
  init(input: I): { ctx: C; entities: E[] };
  /** Entity i at `tick`, from the whole field at tick - 1. Must not read `prev` beyond what it passes in. */
  step(ctx: C, prev: readonly E[], i: number, tick: number): E;
  /** Is the run over after `tick`? A function of the field at that tick only. */
  done(ctx: C, field: readonly E[], tick: number): boolean;
  encode(e: E): Uint8Array;
  decode(b: Uint8Array): E;
}

export interface SteppedRun<C, E> {
  readonly ctx: C;
  /** trajectory[t][i]: entity i after tick t (t = 0 is the initial field). */
  readonly trajectory: E[][];
}

export const defineStepped = <I, C, E>(spec: SteppedSpec<I, C, E>): SteppedSpec<I, C, E> => spec;

export function runStepped<I, C, E>(s: SteppedSpec<I, C, E>, input: I, maxTicks = 1_000_000): SteppedRun<C, E> {
  const { ctx, entities } = s.init(input);
  const trajectory: E[][] = [entities];
  for (let t = 1; t <= maxTicks; t += 1) {
    const prev = trajectory[t - 1]!;
    const next = prev.map((_, i) => s.step(ctx, prev, i, t));
    trajectory.push(next);
    if (s.done(ctx, next, t)) return { ctx, trajectory };
  }
  throw new RangeError(`the run did not finish in ${maxTicks} ticks`);
}

/** sha256 of the field at one tick: every entity's encoding, in order. */
export function fieldHash<E>(s: Pick<SteppedSpec<unknown, unknown, E>, "encode">, field: readonly E[]): Uint8Array {
  const p = createPacker();
  for (const e of field) p.raw(s.encode(e));
  return sha256(p.finish());
}

/** One entity's states across ticks, hashed flat: sha256(encode(s1) || encode(s2) || ...). Encodings are fixed-size, so
 *  the concatenation is unambiguous -- and one pass costs half the blocks of a per-state chain (it's most of a job's work). */
export function traceHash<E>(s: Pick<SteppedSpec<unknown, unknown, E>, "encode">, states: readonly E[]): Uint8Array {
  const p = createPacker();
  for (const e of states) p.raw(s.encode(e));
  return sha256(p.finish());
}

export interface Window { readonly index: number; readonly from: number; readonly to: number }

/** Ticks 1..T cut into windows of `span` ticks ([from, to] inclusive). */
export function windowsOf(lastTick: number, span: number): Window[] {
  const out: Window[] = [];
  for (let from = 1, w = 0; from <= lastTick; from += span, w += 1) out.push({ index: w, from, to: Math.min(lastTick, from + span - 1) });
  return out;
}

/** What a job is handed: the field at from - 1 and every entity's states across the window. */
export interface JobWitness<E> {
  readonly window: Window;
  readonly entity: number;
  readonly boundary: readonly E[];
  /** traces[i][k]: entity i at tick window.from + k. */
  readonly traces: readonly (readonly E[])[];
}

export function jobsOf<C, E>(run: SteppedRun<C, E>, span: number): JobWitness<E>[] {
  return jobsIn(run, windowsOf(run.trajectory.length - 1, span));
}

/**
 * The jobs for windows given outright -- a race cut at its checkpoints, say, rather than every so many ticks. They
 * must tile the run: the first starts at tick 1, each begins where the last ended, and the last ends at the last tick.
 */
export function jobsIn<C, E>(run: SteppedRun<C, E>, windows: readonly Window[]): JobWitness<E>[] {
  const last = run.trajectory.length - 1;
  const n = run.trajectory[0]!.length;
  const jobs: JobWitness<E>[] = [];
  windows.forEach((w, k) => {
    const before = k === 0 ? 1 : windows[k - 1]!.to + 1;
    if (w.index !== k || w.from !== before || w.to < w.from || w.to > last) throw new RangeError("the windows must tile the run, in order");
  });
  if (windows.length && windows[windows.length - 1]!.to !== last) throw new RangeError("the windows must tile the run, in order");
  for (const w of windows) {
    const boundary = run.trajectory[w.from - 1]!;
    const traces = Array.from({ length: n }, (_, i) => run.trajectory.slice(w.from, w.to + 1).map((f) => f[i]!));
    for (let i = 0; i < n; i += 1) jobs.push({ window: w, entity: i, boundary, traces });
  }
  return jobs;
}

/** What a job commits (its public values, beside the program id and the input digest). */
export interface JobOutput<E> {
  readonly window: Window;
  readonly entity: number;
  /** Hash of the witnessed field at from - 1. */
  readonly boundary: Uint8Array;
  /** Hash of every entity's witnessed trace. */
  readonly traces: readonly Uint8Array[];
  /** Hash of this entity's COMPUTED trace. */
  readonly computed: Uint8Array;
  /** This entity's state at the window's last tick (computed). */
  readonly last: E;
}

/** The job program: re-derive entity i across the window from the witnessed field. */
export function proveJob<C, E>(s: SteppedSpec<unknown, C, E>, ctx: C, j: JobWitness<E>): JobOutput<E> {
  const n = j.boundary.length;
  const computed: E[] = [];
  for (let k = 0, t = j.window.from; t <= j.window.to; k += 1, t += 1) {
    const prev = k === 0 ? j.boundary : j.traces.map((tr) => tr[k - 1]!);
    if (prev.length !== n) throw new RangeError("the witness is missing an entity");
    computed.push(s.step(ctx, prev, j.entity, t));
  }
  return {
    window: j.window, entity: j.entity, boundary: fieldHash(s, j.boundary), traces: j.traces.map((tr) => traceHash(s, tr)),
    computed: traceHash(s, computed), last: computed[computed.length - 1]!,
  };
}

/**
 * The aggregator: given every job's output (already proof-verified), check they tile the run and
 * agree, and return the final field. Throws on the first inconsistency.
 */
export function aggregate<I, C, E>(s: SteppedSpec<I, C, E>, input: I, ctx: C, outs: readonly JobOutput<E>[]): { field: E[]; lastTick: number } {
  let field = s.init(input).entities;
  const n = field.length;
  const byWindow = new Map<number, JobOutput<E>[]>();
  for (const o of outs) byWindow.set(o.window.index, [...(byWindow.get(o.window.index) ?? []), o]);
  let expectFrom = 1, lastTick = 0;
  for (let w = 0; byWindow.has(w); w += 1) {
    const jobs = byWindow.get(w)!.slice().sort((a, b) => a.entity - b.entity);
    if (jobs.length !== n || jobs.some((o, i) => o.entity !== i)) throw new RangeError(`window ${w} needs exactly one job per entity`);
    const win = jobs[0]!.window;
    if (win.from !== expectFrom || jobs.some((o) => o.window.from !== win.from || o.window.to !== win.to)) throw new RangeError(`window ${w} does not continue the run`);
    const boundary = fieldHash(s, field);
    for (const o of jobs) {
      if (!equalBytes(o.boundary, boundary)) throw new RangeError(`window ${w}: job ${o.entity} started from another field`);
      if (o.traces.length !== n || o.traces.some((t, i) => !equalBytes(t, jobs[0]!.traces[i]!))) throw new RangeError(`window ${w}: jobs witnessed different traces`);
    }
    jobs.forEach((o, i) => { if (!equalBytes(o.computed, o.traces[i]!)) throw new RangeError(`window ${w}: entity ${i}'s witnessed trace is not what the rules produce`); });
    field = jobs.map((o) => o.last);
    expectFrom = win.to + 1;
    lastTick = win.to;
    const isLast = !byWindow.has(w + 1);
    if (isLast !== s.done(ctx, field, lastTick)) throw new RangeError(isLast ? "the run ends before it is done" : "the run continues after it was done");
  }
  if (lastTick === 0) throw new RangeError("no windows");
  return { field, lastTick };
}

// ---------------------------------------------------------------- the bytes the guests read and commit

/** "HSJW" v1: u32 window · u32 from · u32 to · u32 entity · u32 n · the boundary field · every trace, entity-major. */
export function encodeWitness<E>(s: Pick<SteppedSpec<unknown, unknown, E>, "encode">, j: JobWitness<E>): Uint8Array {
  const p = createPacker().bytes4("HSJW").u8(1).u32(j.window.index).u32(j.window.from).u32(j.window.to).u32(j.entity).u32(j.boundary.length);
  for (const e of j.boundary) p.raw(s.encode(e));
  for (const tr of j.traces) for (const e of tr) p.raw(s.encode(e));
  return p.finish();
}

/** "HSJO" v1: what a job commits; the job's public value is sha256 of these bytes (32 bytes, a Fray chunk statement). */
export function encodeOutput<E>(s: Pick<SteppedSpec<unknown, unknown, E>, "encode">, o: JobOutput<E>): Uint8Array {
  const p = createPacker().bytes4("HSJO").u8(1).u32(o.window.index).u32(o.window.from).u32(o.window.to).u32(o.entity).bytes32(o.boundary).u32(o.traces.length);
  for (const t of o.traces) p.bytes32(t);
  return p.bytes32(o.computed).raw(s.encode(o.last)).finish();
}

File: zk/keel-proof/src/stepped.rs
//! Stepped sims (the Rust twin of @keel-engine/proof stepped.ts): simultaneous ticks, so a run is
//! proven as (entity, window) jobs plus one aggregation. A game implements `Stepped`; the job guest
//! calls `prove_job`, the aggregator guest calls `aggregate`.
use crate::bytes::{Packer, Reader};
use crate::{sha256, Error, Result};
use alloc::vec::Vec;

pub trait Stepped {
    type Ctx;
    type Entity: Clone;
    /// Entity `i` at `tick`, from the whole field at `tick - 1`.
    fn step(&self, ctx: &mut Self::Ctx, prev: &[Self::Entity], i: usize, tick: u32) -> Self::Entity;
    fn done(&self, ctx: &Self::Ctx, field: &[Self::Entity], tick: u32) -> bool;
    fn encode(&self, e: &Self::Entity) -> Vec<u8>;
    fn decode(&self, b: &[u8]) -> Result<Self::Entity>;
    /// Encoded size of one entity (fixed: the witness is a flat array of them).
    const ENTITY_BYTES: usize;
}

pub fn field_hash<S: Stepped>(s: &S, field: &[S::Entity]) -> [u8; 32] {
    let mut p = Packer::new();
    for e in field { p = p.raw(&s.encode(e)); }
    sha256(&p.finish())
}

/// One entity's states hashed flat (fixed-size encodings, so unambiguous): half the blocks of a per-state chain.
pub fn trace_hash<S: Stepped>(s: &S, states: &[S::Entity]) -> [u8; 32] {
    let mut buf = Vec::with_capacity(states.len() * S::ENTITY_BYTES);
    for e in states { buf.extend_from_slice(&s.encode(e)); }
    sha256(&buf)
}

/// A job's witness: the window, the entity, the field at `from - 1`, and every entity's states across the window.
pub struct JobWitness<E> { pub window: u32, pub from: u32, pub to: u32, pub entity: u32, pub boundary: Vec<E>, pub traces: Vec<Vec<E>> }

/// "HSJW" v1: u32 window · u32 from · u32 to · u32 entity · u32 n · boundary n entities · traces n x (to-from+1) entities (entity-major).
pub fn decode_witness<S: Stepped>(s: &S, b: &[u8]) -> Result<JobWitness<S::Entity>> {
    let mut r = Reader::new(b);
    r.magic(b"HSJW")?;
    if r.u8()? != 1 { return Err(Error::Malformed("unknown job witness version")); }
    let (window, from, to, entity, n) = (r.u32()?, r.u32()?, r.u32()?, r.u32()?, r.u32()? as usize);
    if to < from || entity as usize >= n { return Err(Error::Malformed("bad job window")); }
    let len = (to - from + 1) as usize;
    let take = |r: &mut Reader| s.decode(r.take(S::ENTITY_BYTES)?);
    let mut boundary = Vec::with_capacity(n);
    for _ in 0..n { boundary.push(take(&mut r)?); }
    let mut traces = Vec::with_capacity(n);
    for _ in 0..n { let mut t = Vec::with_capacity(len); for _ in 0..len { t.push(take(&mut r)?); } traces.push(t); }
    r.end()?;
    Ok(JobWitness { window, from, to, entity, boundary, traces })
}

pub struct JobOutput<E> { pub window: u32, pub from: u32, pub to: u32, pub entity: u32, pub boundary: [u8; 32], pub traces: Vec<[u8; 32]>, pub computed: [u8; 32], pub last: E }

/// "HSJO" v1: u32 window · u32 from · u32 to · u32 entity · bytes32 boundary · u32 n · n x bytes32 traces · bytes32 computed · last entity.
pub fn encode_output<S: Stepped>(s: &S, o: &JobOutput<S::Entity>) -> Vec<u8> {
    let mut p = Packer::new().bytes4(b"HSJO").u8(1).u32(o.window).u32(o.from).u32(o.to).u32(o.entity).bytes32(&o.boundary).u32(o.traces.len() as u32);
    for t in &o.traces { p = p.bytes32(t); }
    p.bytes32(&o.computed).raw(&s.encode(&o.last)).finish()
}

pub fn decode_output<S: Stepped>(s: &S, b: &[u8]) -> Result<JobOutput<S::Entity>> {
    let mut r = Reader::new(b);
    r.magic(b"HSJO")?;
    if r.u8()? != 1 { return Err(Error::Malformed("unknown job output version")); }
    let (window, from, to, entity, boundary) = (r.u32()?, r.u32()?, r.u32()?, r.u32()?, r.bytes32()?);
    let n = r.u32()? as usize;
    let mut traces = Vec::with_capacity(n);
    for _ in 0..n { traces.push(r.bytes32()?); }
    let computed = r.bytes32()?;
    let last = s.decode(r.take(S::ENTITY_BYTES)?)?;
    r.end()?;
    Ok(JobOutput { window, from, to, entity, boundary, traces, computed, last })
}

/// The job program: re-derive entity `i` across the window from the witnessed field.
pub fn prove_job<S: Stepped>(s: &S, ctx: &mut S::Ctx, j: &JobWitness<S::Entity>) -> Result<JobOutput<S::Entity>> {
    let n = j.boundary.len();
    if j.traces.len() != n { return Err(Error::Witness("the witness is missing an entity")); }
    let mut computed: Vec<S::Entity> = Vec::new();
    let mut prev: Vec<S::Entity> = j.boundary.clone();
    for (k, t) in (j.from..=j.to).enumerate() {
        if k > 0 { prev = j.traces.iter().map(|tr| tr[k - 1].clone()).collect(); }
        computed.push(s.step(ctx, &prev, j.entity as usize, t));
    }
    Ok(JobOutput {
        window: j.window, from: j.from, to: j.to, entity: j.entity,
        boundary: field_hash(s, &j.boundary),
        traces: j.traces.iter().map(|tr| trace_hash(s, tr)).collect(),
        computed: trace_hash(s, &computed),
        last: computed.last().cloned().ok_or(Error::Witness("an empty window"))?,
    })
}

/// The aggregator's checks over every (verified) job output: tiling, agreement, chaining, completion.
/// `start` is the initial field (from the match input). Returns the final field and its tick.
pub fn aggregate<S: Stepped>(s: &S, ctx: &S::Ctx, start: Vec<S::Entity>, mut outs: Vec<JobOutput<S::Entity>>) -> Result<(Vec<S::Entity>, u32)> {
    let n = start.len();
    outs.sort_by(|a, b| a.window.cmp(&b.window).then(a.entity.cmp(&b.entity)));
    if outs.is_empty() || outs.len() % n != 0 { return Err(Error::Run("jobs do not tile the run")); }
    let mut field = start;
    let mut expect_from = 1u32;
    let windows = outs.len() / n;
    for w in 0..windows {
        let jobs = &outs[w * n..(w + 1) * n];
        let (from, to) = (jobs[0].from, jobs[0].to);
        for (i, o) in jobs.iter().enumerate() {
            if o.window != w as u32 || o.entity != i as u32 { return Err(Error::Run("a window needs exactly one job per entity")); }
            if o.from != from || o.to != to || from != expect_from { return Err(Error::Run("a window does not continue the run")); }
            if o.traces.len() != n || o.traces != jobs[0].traces { return Err(Error::Run("jobs witnessed different traces")); }
            if o.computed != o.traces[i] { return Err(Error::Run("a witnessed trace is not what the rules produce")); }
        }
        let boundary = field_hash(s, &field);
        if jobs.iter().any(|o| o.boundary != boundary) { return Err(Error::Run("a job started from another field")); }
        field = jobs.iter().map(|o| o.last.clone()).collect();
        expect_from = to + 1;
        let is_last = w + 1 == windows;
        if is_last != s.done(ctx, &field, to) { return Err(Error::Run(if is_last { "the run ends before it is done" } else { "the run continues after it was done" })); }
    }
    Ok((field, expect_from - 1))
}
